### PR TITLE
Verify and fix some of the RFC-2045 constraints of mimemail

### PR DESCRIPTION
**Pull request by seriyps**
Fixes:
* quoted-printable "very long lines" bug fixed
* never produce bare CR or LF (Fixes: #253)

Validations:
* no bare CR or LF
* no bytes above 127
* no 0-bytes
* no lines longer than ~1000 chars
* quoted-printable should not produce lines over 76 chars

### CODE REVIEW

File: src/mimemail.erl [hunk 1]
```diff
@@ -652,91 +653,98 @@ check_headers([Header | Tail], Headers) ->
 
 ensure_content_headers(Type, SubType, Parameters, Headers, Body, Toplevel) ->
 	CheckHeaders = [<<"Content-Type">>, <<"Content-Disposition">>, <<"Content-Transfer-Encoding">>],
-	ensure_content_headers(CheckHeaders, Type, SubType, Parameters, lists:reverse(Headers), Body, Toplevel).
+	CheckHeadersValues = [{Name, get_header_value(Name, Headers)} || Name <- CheckHeaders],
+	ensure_content_headers(CheckHeadersValues, Type, SubType, Parameters, lists:reverse(Headers), Body, Toplevel).
 
 ensure_content_headers([], _, _, Parameters, Headers, _, _) ->
 	{Parameters, lists:reverse(Headers)};
-ensure_content_headers([Header | Tail], Type, SubType, Parameters, Headers, Body, Toplevel) ->
```
Comment (seriyps) [inline src/mimemail.erl:659]: no actual changes in this function. Just made it less nested.

File: src/mimemail.erl [hunk 2]
```diff
@@ -746,14 +754,52 @@ guess_best_encoding(Body) ->
...
 	if
-		Percent == 100 ->
-			<<"7bit">>;
 		Percent > 80 ->
 			<<"quoted-printable">>;
 		true ->
 			<<"base64">>
 	end.
 
+%% https://tools.ietf.org/html/rfc2045#section-2.7:
```
Comment (seriyps) [inline src/mimemail.erl:763]: now there are much more strict requirements to allow body to be sent as is:

no lines longer than 998 bytes
no bare CR or LF
no 0 bytes or bytes above 127

File: src/mimemail.erl [hunk 3]
```diff
@@ -900,7 +946,7 @@ encode_quoted_printable(Body, Acc, L) when L >= 75 ->
 		SIndex when (L - 75) < SIndex ->
 			% okay, we can safely stick some whitespace in
 			NewAcc = insert_soft_newline(Acc, SIndex - 1),
-			encode_quoted_printable(Body, NewAcc, 0);
+			encode_quoted_printable(Body, NewAcc, SIndex);
 		_ ->
```
Comment (seriyps) [inline src/mimemail.erl:950]: this and next diff are actually bugfixes. Here we used to set "current line length" to 0 even though the line we are building might be much longer than 0.

File: test/prop_mimemail.erl [hunk 4]
```diff
@@ -99,6 +100,80 @@ match({TypeA, SubTypeA, HeadersA, _ParamsA, BodyA},
...
+	   Body,
+	   proper_types:oneof([?SIZED(Size, printable_ascii(Size * 50)),
+						   ?SIZED(Size, printable_ascii_and_cariage(Size * 50)),
+						   printable_ascii(),
+						   printable_ascii_and_cariage(),
+						   nonull_utf8(),
+						   proper_types:binary()]),
+	   begin
+		   [QPEncoded] = mimemail:encode_quoted_printable(Body),
+		   ?assertEqual(Trim(Body), Trim(mimemail:decode_quoted_printable(QPEncoded))),
```
Comment (seriyps) [inline test/prop_mimemail.erl:120]: we still can loose heading and trailing whitespaces and tabs. Not sure if it's a bug, but doesn't sound right. Can be fixed in the following iterations.

**Comment by mworrell** [comment]
Wow, that was quite a bit of work.... good additions, cleanup and fixes.

**Review by mworrell — APPROVED** [review]
Good work, thank you!

**Pull request merged**